plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java

```diff
@@ -167,16 +174,25 @@ public CloseableIterator<List<Object>> iterator()
             addCloseable(planFilesIterator);
 
             return new CloseableIterator<>() {
+                private Iterator<DeleteFile> deleteFileIterator = emptyIterator();
+
                 @Override
                 public boolean hasNext()
                 {
-                    return !closed && planFilesIterator.hasNext();
+                    return !closed && (planFilesIterator.hasNext() || deleteFileIterator.hasNext());
                 }
 
                 @Override
                 public List<Object> next()
                 {
-                    return getRecord(planFilesIterator.next().file());
+                    if (deleteFileIterator.hasNext()) {
```

Comment:
Use `createMetadataTableInstance(icebergTable, FILES)` in `io.trino.plugin.iceberg.FilesTable#cursor` to be able to get delete files from `planFiles()`.

```
        TableScan tableScan = createMetadataTableInstance(icebergTable, FILES)
                .newScan()
                .includeColumnStats();
```

While your solution works as well, I find it retrieves a bit too much data in order to provide the functionality required by `$files` table.

I've done some time ago an experiment related to https://github.com/trinodb/trino/pull/11206 which you may find useful here: 

https://github.com/trinodb/trino/compare/master...findinpath:trino:findinpath/add-all-files

Reply:
Thanks for the detailed info. Let me use the `createMetadataTableInstance` for the table scan.